dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerInterface.java

```diff
@@ -340,90 +347,188 @@ default String getURLFromEntry(String dockerEntry, ToolDAO toolDAO) {
+        String repoUrl = DOCKERHUB_URL + "repositories/" + repo + "/tags?name=" + tagName;
+        DockerHubTag dockerHubTag = new DockerHubTag();
+        do {
+            try {
+                URL url = new URL(repoUrl);
+                response = Optional.of(IOUtils.toString(url, StandardCharsets.UTF_8));
+            } catch (IOException ex) {
+                LOG.error("Unable to get DockerHub response for " + repo);
+                response = Optional.empty();
+            }
+
+            if (response.isPresent()) {
+                Gson gson = new Gson();
+
+                final String json = response.get();
+                errorMap = (Map<String, String>)gson.fromJson(json, errorMap.getClass());
+                if (errorMap.get("message") != null) {
+                    LOG.error("Error response from DockerHub: " + errorMap.get("message"));
+                    return dockerHubImages;
+                }
+
+                // DockerHub seems to give empty results if something is not found, other fields are marked as null
+                dockerHubTag = gson.fromJson(json, DockerHubTag.class);
+                List<Results> results = Arrays.asList(dockerHubTag.getResults());
+                if (results.isEmpty()) {
+                    LOG.error("Could not find any results for " + repo);
+                    break;
+                }
+
+                for (Results r : results) {
+                    if (r.getName().equals(tagName)) {
+                        List<DockerHubImage> images = Arrays.asList(r.getImages());
+                        // For every version, DockerHub can provide multiple images, one for each architecture
+                        images.stream().forEach(dockerHubImage -> {
+                            final String manifestDigest = dockerHubImage.getDigest();
+                            Checksum checksum = new Checksum(manifestDigest.split(":")[0], manifestDigest.split(":")[1]);
+                            List<Checksum> checksums = Collections.singletonList(checksum);
+                            Image archImage = new Image(checksums, repo, tagName, r.getImageID(), Registry.DOCKER_HUB);
+                            archImage.setArchitecture(dockerHubImage.getArchitecture());
+                            dockerHubImages.add(archImage);
+                        });
+                        versionFound = true;
+                        break;
+                    }
+                }
+                if (!versionFound) {
+                    repoUrl = dockerHubTag.getNext();
+                }
+            }
+        } while (response.isPresent() && !versionFound && dockerHubTag.getNext() != null);
+        return dockerHubImages;
+    }
+
+    //default Optional<String> getImageResponseFromQuay(String repo, String tag) {
+    default Set<Image> getImageResponseFromQuay(String repo, String tagName) {
+        final String tagUrl = QUAY_URL + "repository/" + repo + "/tag/" + "?specificTag=" + tagName;
```

Comment:
Would be more readable using swagger-java-quay-client. 
Can take a look at the series of PRs ending with https://github.com/dockstore/dockstore/pull/3273 to demo how to use it.

Reply:
ended up being way easier than I thought.